In [6]:
import numpy as np 
import pandas as pd 
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

**numpy and pandas**: For data manipulation.

**fetch_california_housing**: Loads the dataset.

**train_test_split**: Splits data into training and testing sets.

**StandardScaler**: Normalizes features to improve ANN performance.

**Sequential, Dense**: To build the ANN layers.

**Adam**: Optimizer to update weights during training.

In [ ]:
housing = fetch_california_housing()
x = pd.DataFrame(housing.data, columns = housing.feature_names)
y = housing.target # here x is converted to dataframe because it has many columns, but it is no necessary fro y

In [4]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 42)

In [5]:
scaler = StandardScaler()
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.fit_transform(x_test)

In [10]:
model = Sequential([Input(shape = (x_train_scaled.shape[1],)), # input layer
                    Dense(64, activation = 'relu'), Dense(32, activation = 'relu'), Dense(16, activation = 'relu'), # hidden layers
                    Dense(1)]) # output layer

Sequential model: Layers are stacked one after another.

Dense layers: Fully connected layers.

64 → 32 → 16 neurons in hidden layers.

relu activation introduces non-linearity.

Output layer: Single neuron for predicting a continuous value.

In [14]:
# compile
from sklearn.metrics import mean_absolute_error, mean_squared_error
model.compile(optimizer = Adam(learning_rate = 0.001), loss = 'mse', metrics = ['mae'])

In [15]:
# train the model
history = model.fit(x_train_scaled, y_train, validation_split=0.2, epochs = 50, batch_size = 32)

Epoch 1/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - loss: 1.0622 - mae: 0.6940 - val_loss: 0.4962 - val_mae: 0.5105
Epoch 2/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.4158 - mae: 0.4590 - val_loss: 0.4055 - val_mae: 0.4598
Epoch 3/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3642 - mae: 0.4304 - val_loss: 0.3848 - val_mae: 0.4365
Epoch 4/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - loss: 0.3470 - mae: 0.4156 - val_loss: 0.3635 - val_mae: 0.4211
Epoch 5/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3313 - mae: 0.4065 - val_loss: 0.3481 - val_mae: 0.4150
Epoch 6/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3200 - mae: 0.3969 - val_loss: 0.3467 - val_mae: 0.3991
Epoch 7/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.3093 - mae: 0.3903 - val_loss: 0.3418 - val_mae: 0.4145
Epoch 8/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - loss: 0.3010 - mae: 0.3830 - val_loss: 0.3263 - val_mae: 0.3964
Epoch 9/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - lo

validation_split=0.2: 20% of training data used to validate model during training.

epochs=50: Model sees the entire training set 50 times.

batch_size=32: Number of samples per weight update.

In [17]:
# evaluate the model
loss, mae = model.evaluate(x_test_scaled, y_test)
print(f'Test MAE:{mae}')

129/129 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 3.5743 - mae: 1.2621
Test MAE:1.262130856513977


Test MAE: 1.26 → On average, the predicted house prices differ from the actual prices by 1.262 units (in the dataset’s scale).

Model performance: The ANN predicts house prices reasonably well, capturing the relationship between features and prices.

In [19]:
# make predictions
predictions = model.predict(x_test_scaled)
print('first 2 predictions:', predictions[:5])

129/129 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step
first 2 predictions: [[0.59323215]
 [2.696172  ]
 [9.822733  ]
 [2.2936037 ]
 [6.0885534 ]]


In [23]:
comparison = pd.DataFrame({'Actual': y_test[:10]# here it is already 1 d, so no need.
                           , 'Predicted': predictions[:10].flatten()})
print(comparison) # predictions[:10].flatten() - is used to So it changes the predictions from a 2D array into a 1D array.

    Actual  Predicted
0  0.47700   0.593232
1  0.45800   2.696172
2  5.00001   9.822733
3  2.18600   2.293604
4  2.78000   6.088553
5  1.58700   1.802329
6  1.98200   2.026820
7  1.57500   1.534362
8  3.40000   4.016091
9  4.46600   4.618655
